src/MonitorProfiler/EventProvider/ProfilerEvent.h

```diff
@@ -0,0 +1,179 @@
+    return Initialize<index + 1, TArgs...>(names);
+}
+
+template<typename... Args>
+template<size_t index>
+HRESULT ProfilerEvent<Args...>::Initialize(const WCHAR* (&names)[sizeof...(Args)])
+{
+    //CONSIDER An alternate design is to make each event call DefineEvent instead of leaving this responsibility on the provider.
+    return S_OK;
+}
+
+template<typename... Args>
+HRESULT ProfilerEvent<Args...>::WritePayload(const Args&... args)
+{
+    COR_PRF_EVENT_DATA data[sizeof...(Args)];
+    return WritePayload<0, Args...>(data, args...);
+}
+
+template<typename... Args>
+template<size_t index, typename T, typename... TArgs>
+HRESULT ProfilerEvent<Args...>::WritePayload(COR_PRF_EVENT_DATA* data, const T& first, TArgs... rest)
+{
+    data[index].ptr = reinterpret_cast<UINT64>(&first);
+    data[index].size = static_cast<UINT32>(sizeof(T));
+    data[index].reserved = 0;
+    return WritePayload<index + 1, TArgs...>(data, rest...);
+}
+
+template<typename... Args>
+template<size_t index, typename T, typename... TArgs>
+HRESULT ProfilerEvent<Args...>::WritePayload(COR_PRF_EVENT_DATA* data, const tstring& first, TArgs... rest)
+{
+    if (first.size() == 0)
+    {
+        data[index].ptr = 0;
+        data[index].size = 0;
+        data[index].reserved = 0;
+    }
+    else
+    {
+        data[index].ptr = reinterpret_cast<UINT64>(first.c_str());
+        data[index].size = static_cast<UINT32>((first.size() + 1) * sizeof(WCHAR));
+        data[index].reserved = 0;
+    }
+    return WritePayload<index + 1, TArgs...>(data, rest...);
+}
+
+ template<typename... Args>
+ template<size_t index, typename T, typename... TArgs>
+ HRESULT ProfilerEvent<Args...>::WritePayload(COR_PRF_EVENT_DATA* data, const std::vector<typename T::value_type>& first, TArgs... rest)
+ {
+     std::vector<BYTE> buffer(0);
+
+     if (first.size() == 0)
+     {
+         data[index].ptr = 0;
+         data[index].size = 0;
+         data[index].reserved = 0;
+     }
+     else
+     {
+         buffer = std::move(GetEventBuffer(first));
+         data[index].ptr = reinterpret_cast<UINT64>(buffer.data());
+         data[index].size = static_cast<UINT32>(buffer.size());
+         data[index].reserved = 0;
+     }
+     return WritePayload<index + 1, TArgs...>(data, rest...);
+ }
+
+template<typename... Args>
+template<typename T>
+std::vector<BYTE> ProfilerEvent<Args...>::GetEventBuffer(const std::vector<T>& data)
+{
+    size_t offset = 0;
+    size_t bufferSize = 2 + (data.size() * sizeof(T));
+    std::vector<BYTE> buffer = std::vector<BYTE>(bufferSize);
+    WriteToBuffer<UINT16>(buffer.data(), bufferSize, &offset, (UINT16)data.size());
+
+    for (const T& element : data)
+    {
+        WriteToBuffer<T>(buffer.data(), bufferSize, &offset, element);
+    }
+
+    return buffer;
+}
+
+template<typename... Args>
+template<typename T>
+void ProfilerEvent<Args...>::WriteToBuffer(BYTE* pBuffer, size_t bufferLength, size_t* pOffset, const T& value)
```

Comment:
It doesn't appear that we're using `bufferLength`; should we be? Or can this be removed from the signature?